numpyro/infer/autoguide.py

```diff
@@ -213,6 +213,103 @@ def quantiles(self, params, quantiles):
         raise NotImplementedError
 
 
+class AutoGuideList(AutoGuide):
+    """
+    Container class to combine multiple automatic guides.
+
+    Example usage::
+
+        rng_key_init = random.PRNGKey(0)
+        guide = AutoGuideList(my_model)
+        guide.append(
+            AutoNormal(
+                numpyro.handlers.block(numpyro.handlers.seed(model, rng_seed=0), hide=["coefs"])
+            )
+        )
+        guide.append(
+            AutoDelta(
+                numpyro.handlers.block(numpyro.handlers.seed(model, rng_seed=1), expose=["coefs"])
+            )
+        )
+        svi = SVI(model, guide, optim, Trace_ELBO())
+        svi_state = svi.init(rng_key_init, data, labels)
+        params = svi.get_params(svi_state)
+
+    :param callable model: a NumPyro model
+    """
+
+    def __init__(
+        self, model, *, prefix="auto", init_loc_fn=init_to_uniform, create_plates=None
+    ):
+        self._guides = []
+        super().__init__(
+            model, prefix=prefix, init_loc_fn=init_loc_fn, create_plates=create_plates
+        )
+
+    def append(self, part):
+        """
+        Add an automatic or custom guide for part of the model. The guide should
+        have been created by blocking the model to restrict to a subset of
+        sample sites. No two parts should operate on any one sample site.
+
+        :param part: a partial guide to add
+        :type part: AutoGuide
+        """
+        self._guides.append(part)
+
+    def __call__(self, *args, **kwargs):
+        if self.prototype_trace is None:
+            # run model to inspect the model structure
+            self._setup_prototype(*args, **kwargs)
+
+        # create all plates
+        self._create_plates(*args, **kwargs)
+
+        # run slave guides
+        result = {}
+        for part in self._guides:
+            result.update(part(*args, **kwargs))
+        return result
+
+    def __getitem__(self, key):
+        return self._guides[key]
+
+    def __len__(self):
+        return len(self._guides)
+
+    def __iter__(self):
+        yield from self._guides
+
+    def sample_posterior(self, rng_key, params, *args, sample_shape=(), **kwargs):
+        result = {}
+        for part in self._guides:
+            if isinstance(part, numpyro.infer.autoguide.AutoDelta) or isinstance(
+                part, numpyro.infer.autoguide.AutoSemiDAIS
+            ):
+                result.update(
+                    part.sample_posterior(
+                        rng_key, params, *args, sample_shape=sample_shape, **kwargs
+                    )
+                )
+            else:
+                result.update(
+                    part.sample_posterior(rng_key, params, sample_shape=sample_shape)
```

Comment:
Could you usage signature of other algorithm to be the same as AutoDelta? That signature is more general I think.

Reply:
To reduce the chance of misunderstanding, can you please clarify what you mean.

The reason for the if-else clause is that `AutoDelta` and `AutoSemiDAIS` differ from the other autoguides when it comes to the signature of `sample_posterior()`:
- `AutoDelta` and `AutoSemiDAIS` have `sample_posterior(self, rng_key, params, *args, sample_shape=(), **kwargs)`
- other autoguides have `sample_posterior(self, rng_key, params, sample_shape=())`

I suppose we could break things if we change the signature of `sample_posterior()` since `sample_shape` might have been passed as a positional argument.

Reply:
> since sample_shape might have been passed as a positional argument

Good point! Maybe we can keep your current implementation here as-is and adding `*,` before sample_shape in other classes. After the next release, we'll make the signature consistent. Users of the next release will keep informed of the change because they are no longer able to use `sample_shape` as a positional argument.

Reply:
Sounds reasonable to me! I'm happy to create another PR to address this if you want.

Reply:
Yes, thank you!